File: src/TypeFormatter/ArrayTypeFormatter.ts
import type { Definition } from "../Schema/Definition.js";
import type { SubTypeFormatter } from "../SubTypeFormatter.js";
import { ArrayType } from "../Type/ArrayType.js";
import type { BaseType } from "../Type/BaseType.js";
import type { GetDefinitionOptions } from "../TypeFormatter.js";
import type { TypeFormatter } from "../TypeFormatter.js";

export class ArrayTypeFormatter implements SubTypeFormatter {
    public constructor(private childTypeFormatter: TypeFormatter) {}

    public supportsType(type: BaseType): boolean {
        return type instanceof ArrayType;
    }
    public getDefinition(type: ArrayType, options?: GetDefinitionOptions): Definition {
        return {
            type: "array",
            items: this.childTypeFormatter.getDefinition(type.getItem(), options),
        };
    }
    public getChildren(type: ArrayType): BaseType[] {
        return this.childTypeFormatter.getChildren(type.getItem());
    }
}

File: internal/formatter/array_type_formatter.go
package formatter

import (
	"github.com/vega/ts-json-schema-generator/internal/schema"
	"github.com/vega/ts-json-schema-generator/internal/types"
)

// ArrayTypeFormatter mirrors src/TypeFormatter/ArrayTypeFormatter.ts.
type ArrayTypeFormatter struct {
	childTypeFormatter TypeFormatter
}

func NewArrayTypeFormatter(childTypeFormatter TypeFormatter) *ArrayTypeFormatter {
	return &ArrayTypeFormatter{childTypeFormatter: childTypeFormatter}
}

func (f *ArrayTypeFormatter) SupportsType(t types.Type) bool {
	_, ok := t.(*types.ArrayType)
	return ok
}

func (f *ArrayTypeFormatter) GetDefinition(t types.Type) *schema.Definition {
	return &schema.Definition{
		Type:  "array",
		Items: f.childTypeFormatter.GetDefinition(t.(*types.ArrayType).Item),
	}
}

func (f *ArrayTypeFormatter) GetChildren(t types.Type) []types.Type {
	return f.childTypeFormatter.GetChildren(t.(*types.ArrayType).Item)
}
